### ISSUE TITLE
Regular crashes of logspout (with logstash adapter)

### ISSUE THREAD

**Issue by paswer** [gliderlabs/logspout#274]
Hi there, 

I have logspout running as a feeder for ELK. It works, meaning I see logs in Kibana. It's just that it regularly (sometimes every couple of seconds) crashes. The other ELK containers (it's all running on docker) continue to work fine.

It's basically the custom version of gliderlabs/logspout:master (https://github.com/gliderlabs/logspout/tree/master/custom), just the modules.go looks like this to enable the forward to logstash:

```
package main

import (
  _ "github.com/looplab/logspout-logstash"
  _ "github.com/gliderlabs/logspout/transports/tcp"
)
```

logspout gets started with this setting: ROUTE_URIS=logstash+tcp://logstash:5000

One thing we found out is that logspout crashes when you log a null (as a string). You can reproduce this by logging a null:
`docker exec [hash-of-any-container-logstash-is-connected-to] bash -c 'echo "null" > /proc/1/fd/1'`

That's the log of a crashed logspout:

> panic: assignment to entry in nil map
> 
> goroutine 21 [running]:
> panic(0x5638366bc2e0, 0xc42052e8c0)
> 	/usr/lib/go/src/runtime/panic.go:500 +0x1a5
> github.com/looplab/logspout-logstash.(*LogstashAdapter).Stream(0xc420106390, 0xc420100300)
> 	/go/src/github.com/looplab/logspout-logstash/logstash.go:143 +0x47d
> github.com/gliderlabs/logspout/router.(*RouteManager).route(0xc4200554c0, 0xc4200a6580)
> 	/go/src/github.com/gliderlabs/logspout/router/routes.go:147 +0xb5
> github.com/gliderlabs/logspout/router.(*RouteManager).Run.func1(0xc4200554c0, 0xc4200a6580)
> 	/go/src/github.com/gliderlabs/logspout/router/routes.go:170 +0x37
> created by github.com/gliderlabs/logspout/router.(*RouteManager).Run
> 	/go/src/github.com/gliderlabs/logspout/router/routes.go:172 +0xf2

Did someone encounter something similar? Is this rather a bug or a wrong usage?

**Comment by mattaitchison** [comment]
It looks to me like a bug in the logstash adapter. 

What commit of logspout-logstash are you using?

**Comment by zconnelly** [comment]
I agree that it looks like it's coming from the logstash adapter.

I'm not terribly familiar with `go` but it seems like something along the lines of `map[nil]["foo"] = "bar"` is happening.

I have a suspicion that it's using `json.Unmarshal` on the message and getting `nil` back (from null) (see: https://golang.org/pkg/encoding/json/#Unmarshal)

**TLDR** - I think we should close this and take it over to https://github.com/looplab/logspout-logstash/

**Comment by zconnelly** [comment]
I am nearly certain that the bug is here: https://github.com/looplab/logspout-logstash/blob/master/logstash.go#L134

```
package main;

import (
    "encoding/json"
    "fmt"
)

func main() {
    var data map[string]interface{}
    var err error
    m_data := []byte(`null`)

    if err = json.Unmarshal([]byte(m_data), &data); err != nil {
        fmt.Println("This doesn't get executed because no error is thrown")
    }

    data["docker"] = `blah`

}
```

The above `go` snippet demonstrates what is happening. The message of `null` is decoded to `nil` and then the `data["docker"] = \`blah\`` causes the panic

```
zacmimi@macbook ~/projects/null_bug $ ./null_bug
panic: assignment to entry in nil map

goroutine 1 [running]:
main.main()
        /Users/zacmimi/projects/null_bug/null_bug.go:17 +0x109
```

I'll submit a PR to https://github.com/looplab/logspout-logstash

**Comment by paswer** [comment]
zconnelly's fix actually solved it. Thanks a lot! I hope, the PR gets integrated soon.
